packages/flutter/lib/src/widgets/shape_boundary.dart

```diff
@@ -0,0 +1,142 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ui';
+
+import 'framework.dart';
+
+/// An abstract class that defines the boundaries of a geometric shape.
+///
+/// [isWithinBoundary] Returns whether the specified shape is within the boundary
+///
+/// [nearestShapeWithinBoundary] Returns the shape closest to the specified shape
+/// within the boundary, or returns the shape as is if it is within the boundary.
+///
+/// See also:
+/// * [OffsetBoundaryProvider], A widget that provides the [Offset] of a [ShapeBoundary]
+/// * [RectBoundaryProvider], A widget that provides the [Rect] of a [ShapeBoundary]
+abstract class ShapeBoundary<T> {
+  /// Returns whether the specified shape is within the boundary.
+  bool isWithinBoundary(T shape);
+
+  /// Returns the shape closest to the specified shape within the boundary.
+  /// If the specified shape is within the boundary, it returns the shape as is.
+  T nearestShapeWithinBoundary(T shape);
+}
+
+/// This widget can provide a [ShapeBoundary] of [Offset] to its descendants,
+/// whose bounds is the current position of this widget.
+///
+/// [ShapeBoundary.isWithinBoundary] returns whether the specified [Offset]
+/// is within the range of this widget. [Offset] should be specified in global coordinates
+///
+/// [ShapeBoundary.nearestShapeWithinBoundary] returns the [Offset] closest to the specified [Offset]
+/// within the boundary, or returns the [Offset] as is if it is within the boundary.
+class OffsetBoundaryProvider extends InheritedWidget {
```

Comment:
Will this be used in the framework at all, i.e. is it necessary to accomplish https://github.com/flutter/flutter/pull/146182 ?

Reply:
Yes, if merged, `RectBoundaryProvider` will be used to provide boundaries for `SliverReorderableList`.